src/repo/remote.rs

```diff
@@ -0,0 +1,692 @@
+        url,
+        Duration::from_secs(15),
+        REMOTE_INDEX_SIZE_LIMIT,
+        date_modified,
+    )?;
+
+    if let Some(pack_index_bytes) = pack_index_bytes {
+        if let Err(e) = PackIndex::parse(pack_index_bytes.as_slice()) {
+            log::error!(
+                "Failed to fetch {} from remote: The remote returned a broken .pack.idx! {}",
+                url,
+                e
+            );
+        } else {
+            log::info!(
+                "Writing {} ({} B)...",
+                pack_index_path.display(),
+                pack_index_bytes.len()
+            );
+            fs::write(&pack_index_path, pack_index_bytes.as_slice()).map_err(Error::IOError)?;
+        }
+    }
+    Ok(())
+}
+
+/// Fetches the remote index from the server.
+pub fn fetch_remote(url: &str, path: &Path) -> Result<RemoteIndex, Error> {
+    let url = url.parse::<Url>().unwrap();
+    let response =
+        read_remote_resource(&url, Duration::from_secs(15), REMOTE_INDEX_SIZE_LIMIT, None)?;
+
+    match response {
+        None => unreachable!(),
+        Some(data) => {
+            let mut remote = RemoteIndex::read(BufReader::new(data.as_slice())).reify(url)?;
+            // Update the .esi
+            create_file(path)
+                .map_err(Error::IOError)?
+                .write_all(data.as_slice())
+                .map_err(Error::IOError)?;
+            // And return the parsed index
+            remote.path = Some(path.to_owned());
+            Ok(remote)
+        }
+    }
+}
+
+/// Fetches the new newest version of the [`RemoteIndex`] from the server and
+/// overwrites its backing file only when the remote file is newer.
+pub fn update_remote(remote: &RemoteIndex) -> Result<RemoteIndex, Error> {
+    let path = remote
+        .path
+        .as_ref()
+        .expect("The RemoteIndex must have a valid .path set! Use RemoteIndex::load().");
+
+    // Read the modification date of the .esi.
+    let date_modified = fs::metadata(path).ok().and_then(|x| x.modified().ok());
+    let url = remote.url.parse::<Url>().unwrap();
+    let response = read_remote_resource(
+        &url,
+        Duration::from_secs(15),
+        REMOTE_INDEX_SIZE_LIMIT,
+        date_modified,
+    )?;
+
+    match response {
+        // The local version is up-to-date.
+        None => RemoteIndex::load(&path).reify(path.display()),
+        Some(data) => {
+            let mut remote = RemoteIndex::read(BufReader::new(data.as_slice())).reify(url)?;
+            // Update the .esi
+            create_file(path)
+                .map_err(Error::IOError)?
+                .write_all(data.as_slice())
+                .map_err(Error::IOError)?;
+            // And return the parsed index
+            remote.path = Some(path.clone());
+            Ok(remote)
+        }
+    }
+}
+
+/// A convenience function which verifies the checksum of the file
+/// and coerces all errors (including ENOENT) to false.
```

Comment:
Hiding IO errors can be confusing, if they ever happen. Fair enough if you want to hide ENOENT, perhaps, but other errors should probably at least be logged, otherwise someone might have quite an unfun day debugging an invisible error.

Reply:
Ack.